### DISCUSSION TITLE
Set in `params` treated differently as List

### DISCUSSION THREAD

**Discussion by finswimmer** [encode/httpx#2847] [Potential Issue] [answered]
Hey,

if one of my params contains a list, `httpx` happily repeat the parameter in the URL for each value.

```py
httpx.Request('GET', 'http://example.com', params={"a": [1, 2 ,3]}).url
# URL('http://example.com?a=1&a=2&a=3')
```

I would expect this behavior as well, if the parameter contains a set. But instead the set is turned into a string.

```py
httpx.Request('GET', 'http://example.com', params={"a": {1, 2 ,3}}).url
# URL('http://example.com?a=%7B1%2C%202%2C%203%7D')
```

Is this a feature or a bug? :smiley: 

I'm using httpx version `0.25.0`.

fin swimmer

**Comment by lovelydinosaur** [comment]
> Is this a feature or a bug?

Neither, it's Python.

**Reply by finswimmer** [reply]
> Neither, it's Python.

Hm, can you please add some more information to this?

**Reply by karpetrosyan** [reply, marked as answer]
You've used an unsupported type, and you're getting unexpected results. If you use an IDE tool or mypy for Python typing, it should probably warn you that the type you are using is incorrect.

type you have used for params is: typing.Dict[str, set[int]]

Supported types:

``` python
QueryParamTypes = Union[
    "QueryParams",
    Mapping[str, Union[PrimitiveData, Sequence[PrimitiveData]]],
    List[Tuple[str, PrimitiveData]],
    Tuple[Tuple[str, PrimitiveData], ...],
    str,
    bytes,
] 
```

Valid examples:

```python

[("test", "test"), ] #  valid for List[Tuple[str, PrimitiveData]]
{"test": "test} # valid for Mapping[str, Union[PrimitiveData, Sequence[PrimitiveData]]]

```

**Reply by lovelydinosaur** [reply]
I'm just being whimsical.
I mean it's not a strongly typed language, and behaviours for unexpected types are often a bit under-defined.

Would {i, you, one} expect a set here to?...

* be coerced into a str, as the default behaviour for any unexpected types.
* be coerced into a list, as the default behaviour for any iterables.
* raise a type error, because we're being strict about the allowed types here.

**Reply by karpetrosyan** [reply]
It's more natural to let Python work as it should. It will be very complex for us to track these kinds of cases.

**Reply by finswimmer** [reply]
Thanks for clarification :+1: Makes sense.

> be coerced into a list, as the default behavior for any iterables.

This was my expectation. I stumbled about this issue, while migrating my code from `requests` to `httpx` and it used to work before. So you can take this as a feature request :smiley:
